core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c

```diff
@@ -16,87 +18,175 @@
+static __maybe_unused void disable_io_compensation(int cmpcr_offset)
 {
-	vaddr_t syscfg_base = get_syscfg_base();
-	uint64_t timeout_ref = 0;
+	vaddr_t cmpcr_base = get_syscfg_base() + cmpcr_offset;
+	uint32_t value_cmpcr = 0;
+	uint32_t apsrc_ansrc = 0;
+	uint32_t value_cmpcr2 = 0;
+
+	value_cmpcr = io_read32(cmpcr_base);
+	value_cmpcr2 = io_read32(cmpcr_base + CMPENSETR_OFFSET);
+	if (!(value_cmpcr & SYSCFG_CMPCR_READY &&
+	      value_cmpcr2 & SYSCFG_CMPENSETR_MPU_EN))
+		return;
+
+	/* Copy APSRC (resp. ANSRC) in RAPSRC (resp. RANSRC) */
+	apsrc_ansrc = value_cmpcr >> SYSCFG_CMPCR_ANSRC_SHIFT;
+	value_cmpcr &= ~(SYSCFG_CMPCR_RANSRC | SYSCFG_CMPCR_RAPSRC);
+	value_cmpcr |= SHIFT_U32(apsrc_ansrc, SYSCFG_CMPCR_RANSRC_SHIFT);
+
+	io_write32(cmpcr_base, value_cmpcr | SYSCFG_CMPCR_SW_CTRL);
 
+	io_setbits32(cmpcr_base + CMPENCLRR_OFFSET, SYSCFG_CMPENSETR_MPU_EN);
+
+	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(cmpcr_base));
+}
+
+static TEE_Result stm32mp1_iocomp(void)
+{
 	if (clk_enable(stm32mp_rcc_clock_id_to_clk(CK_CSI)) ||
 	    clk_enable(stm32mp_rcc_clock_id_to_clk(SYSCFG)))
 		panic();
 
-	io_setbits32(syscfg_base + SYSCFG_CMPENSETR, SYSCFG_CMPENSETR_MPU_EN);
+	enable_io_compensation(SYSCFG_CMPCR);
 
-	timeout_ref = timeout_init_us(SYSCFG_CMPCR_READY_TIMEOUT_US);
+	return TEE_SUCCESS;
+}
 
-	while (!(io_read32(syscfg_base + SYSCFG_CMPCR) & SYSCFG_CMPCR_READY)) {
-		if (timeout_elapsed(timeout_ref)) {
-			EMSG("IO compensation cell not ready");
-			/* Allow an almost silent failure here */
-			break;
-		}
-	}
+driver_init(stm32mp1_iocomp);
 
-	io_clrbits32(syscfg_base + SYSCFG_CMPCR, SYSCFG_CMPCR_SW_CTRL);
+#ifdef CFG_STM32MP13
+void stm32mp_set_vddsd_comp_state(enum stm32mp13_vddsd_comp_id id, bool enable)
+{
+	int cmpcr_offset = 0;
+
+	switch (id) {
+	case SYSCFG_IO_COMP_IDX_SD1:
+		cmpcr_offset = SYSCFG_CMPSD1CR;
+		break;
+	case SYSCFG_IO_COMP_IDX_SD2:
+		cmpcr_offset = SYSCFG_CMPSD2CR;
+		break;
+	default:
+		panic();
+	}
 
-	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(syscfg_base + SYSCFG_CMPCR));
+	if (enable)
+		enable_io_compensation(cmpcr_offset);
+	else
+		disable_io_compensation(cmpcr_offset);
 }
 
-void stm32mp_syscfg_disable_io_compensation(void)
+void stm32mp_set_hslv_state(enum stm32mp13_hslv_id id, bool enable)
 {
-	vaddr_t syscfg_base = get_syscfg_base();
-	uint32_t value = 0;
+	size_t hslvenxr_offset = 0;
+	uint32_t hlvs_value = 0;
 
-	value = io_read32(syscfg_base + SYSCFG_CMPCR) >>
-		SYSCFG_CMPCR_ANSRC_SHIFT;
+	assert(id < SYSCFG_HSLV_COUNT);
 
-	io_clrbits32(syscfg_base + SYSCFG_CMPCR,
-		     SYSCFG_CMPCR_RANSRC | SYSCFG_CMPCR_RAPSRC);
+	if (enable)
+		hlvs_value = SYSCFG_HSLV_KEY;
 
-	value = io_read32(syscfg_base + SYSCFG_CMPCR) |
-		(value << SYSCFG_CMPCR_RANSRC_SHIFT);
+	/* IDs are indices of SYSCFG_HSLVENxR registers */
+	hslvenxr_offset = SYSCFG_HSLVEN0R + id * sizeof(uint32_t);
 
-	io_write32(syscfg_base + SYSCFG_CMPCR, value | SYSCFG_CMPCR_SW_CTRL);
+	io_write32(get_syscfg_base() + hslvenxr_offset, hlvs_value);
```

Comment:
We  could read back if High speed mode is enabled by reading 1 to this register

`
	if (enable && !(io_write32(get_syscfg_base() + hslvenxr_offset) & 0x1))
		panic();
	else if (!enable && io_write32(get_syscfg_base() + hslvenxr_offset) )
		panic();
`